src/transformers/trainer.py

```diff
@@ -2787,17 +2787,17 @@ def save_model(self, output_dir: Optional[str] = None, _internal_call: bool = Fa
                 raise ValueError("Install Accelerate from main branch")
             try:
                 state_dict = self.accelerator.get_state_dict(self.deepspeed)
-                if self.args.should_save:
-                    self._save(output_dir, state_dict=state_dict)
             except ValueError:
+                state_dict = {}
                 logger.warning(
                     " stage3_gather_16bit_weights_on_model_save=false. Saving the full checkpoint instead, use"
                     " zero_to_fp32.py to recover weights"
                 )
-                self._save(output_dir, state_dict={})
-                # remove the dummy state_dict
-                remove_dummy_checkpoint(self.args.should_save, output_dir, [WEIGHTS_NAME, SAFE_WEIGHTS_NAME])
-                self.model_wrapped.save_checkpoint(output_dir)
+            if self.args.should_save:
+                self._save(output_dir, state_dict=state_dict)
+            # remove the dummy state_dict
+            remove_dummy_checkpoint(self.args.should_save, output_dir, [WEIGHTS_NAME, SAFE_WEIGHTS_NAME])
```

Comment:
This would remove the legitimate model checkpoint when `stage3_gather_16bit_weights_on_model_save=True`. `remove_dummy_checkpoint` should only be called in the exception block.